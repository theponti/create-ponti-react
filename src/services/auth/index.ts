/* eslint-disable no-param-reassign */
import { createAsyncThunk, createSlice, PayloadAction } from "@reduxjs/toolkit";
import { Session } from "@supabase/supabase-js";
import { useState } from "react";
import { useNavigate } from "react-router-dom";
import { createUser, getUserById, supabase } from "services/supabase";

export interface AuthState {
  authenticateError?: boolean;
  isLoadingAuth: boolean;
  logoutError?: boolean;
  session: Session | null;
  user?: User;
}

const initialState: AuthState = {
  isLoadingAuth: true,
  session: null,
};

export function useDeleteUser() {
  const navigate = useNavigate();
  const [loading, setLoading] = useState(false);
  const [error, setError] = useState<string | null>(null);

  async function deleteUser(id: string) {
    try {
      const { error: deleteUserError } = await supabase.auth.admin.deleteUser(
        id
      );

      if (deleteUserError) {
        setError(deleteUserError.message);
        return;
      }

      // Delete user profile
      const { error: profileDeleteError } = await supabase
        .from("profiles")
        .delete()
        .match({ id });

      if (profileDeleteError) {
        setError(profileDeleteError.message);
        return;
      }

      navigate("/");
    } catch (err: any) {
      setError(err.message);
    } finally {
      setLoading(false);
    }
  }

  return { deleteUser, error, loading };
}

export const setSession = createAsyncThunk(
  "auth/set-session",
  async (session: Session | null) => {
    if (session) {
      const { data: existingUser, error } = await getUserById(session.user.id);

      // Create new user if none exists
      if (!existingUser || error?.code === "PGRST116") {
        const { data } = await createUser({
          email: session.user.email as string,
          id: session.user.id as string,
        });
        return { session, user: data };
      }

      return { session, user: existingUser };
    }

    return { session };
  }
);

export const authSlice = createSlice({
  name: "auth",
  initialState,
  // The `reducers` field lets us define reducers and generate associated actions
  reducers: {
    setUser(state, action: PayloadAction<User | undefined>) {
      state.isLoadingAuth = false;
      state.user = action.payload;
      if (action.payload === undefined) {
        state.session = null;
      }
    },
  },
  // The `extraReducers` field lets the slice handle actions defined elsewhere,
  // including actions generated by createAsyncThunk or in other slices.
  extraReducers(builder) {
    builder.addCase(setSession.fulfilled, (state, { payload }) => {
      state.user = payload.user;
      state.session = payload.session;
      state.isLoadingAuth = false;
    });
  },
});

export const { setUser } = authSlice.actions;
export default authSlice.reducer;
