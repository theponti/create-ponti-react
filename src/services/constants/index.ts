export { default as ROUTES } from './routes';

export const APP_NAME = 'Ponti UI';
