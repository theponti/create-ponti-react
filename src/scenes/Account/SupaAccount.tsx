import { useCallback, useState } from 'react';
import { Navigate } from 'react-router-dom';

import Button from 'components/Button';
import { supabase } from 'services/supabase';

function SupaAccount({ user }: { user: User }) {
  const [loading, setLoading] = useState(false);
  const [name, setName] = useState(null);

  const onUpdateProfile = useCallback(async (e) => {
    e.preventDefault();

    try {
      setLoading(true);

      const updates = {
        id: user.id,
        name,
        updated_at: new Date(),
      };

      const { error } = await supabase.from('profiles').upsert(updates);

      if (error) {
        throw (error as any);
      }
    } catch (error: any) {
      alert(error.message);
    } finally {
      setLoading(false);
    }
  }, [user, name]);

  const onSetUsername = useCallback((e) => setName(e.target.value), []);

  if (!user) {
    return <Navigate to="/signin" />;
  }

  return (
    <div aria-live="polite">
      {loading ? (
        'Saving ...'
      ) : (
        <form onSubmit={onUpdateProfile} className="form-widget">
          <div>
            Email:
            {user.email}
          </div>
          <div>
            <label htmlFor="name">
              Name
              <input
                id="name"
                type="text"
                value={name || ''}
                onChange={onSetUsername}
              />
            </label>
          </div>
          <div>
            <Button className="button primary block" disabled={loading} type="submit" variant="success">
              Update profile
            </Button>
          </div>
        </form>
      )}
    </div>
  );
}

export default SupaAccount;
