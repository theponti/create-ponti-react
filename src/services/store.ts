import { Action, configureStore, ThunkAction } from '@reduxjs/toolkit';
import auth from './auth';

export const store = configureStore({
  reducer: {
    auth,
  },
});

export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof store.getState>;
export type AppThunk<ReturnType = void> = ThunkAction<
ReturnType,
RootState,
unknown,
Action<string>
>;

// The function below is called a selector and allows us to select a value from
// the state. Selectors can also be defined inline where they're used instead of
// in the slice file. For example: `useSelector((state: RootState) => state.auth.value)`
export const authSelectors = {
  getAuthenticateError: (state: RootState) => state.auth.authenticateError,
  getUser: (state: RootState) => state.auth.user,
  getLoginEmail: (state: RootState) => state.auth.loginEmail,
};
