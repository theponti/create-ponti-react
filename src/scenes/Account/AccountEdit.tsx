import { useCallback, useState } from "react";
import { connect } from "react-redux";
import { useNavigate } from "react-router-dom";

import AlertError from "components/Alerts/Error";
import LoadingButton from "components/Buttons/LoadingButton";
import { setUser } from "services/auth";
import { ACCOUNT_PATH } from "services/constants/routes";
import { AppState, authSelectors, store } from "services/store";
import { useAccountEdit } from "services/supabase";

function AccountEdit({ user }: { user: User }) {
  const navigate = useNavigate();
  const [name, setName] = useState(user.name);
  const { editAccount, error, loading } = useAccountEdit({ userId: user.id });
  const onSetUsername = useCallback((e: any) => setName(e.target.value), []);
  const onFormSubmit = useCallback(
    async (e: any) => {
      e.preventDefault();
      await editAccount({ name });
      store.dispatch(setUser({ ...user, name }));
      navigate(ACCOUNT_PATH);
    },
    [editAccount, navigate, name, user]
  );

  return (
    <div aria-live="polite">
      {error ? (
        <AlertError message="There was an issue updating your profile." />
      ) : null}

      <form onSubmit={onFormSubmit} className="form-widget">
        <div className="form-control mb-8">
          <label htmlFor="name" className="input-group">
            <span>Name</span>
            <input
              id="name"
              type="text"
              className="input input-md input-bordered w-full max-w-xs"
              value={name || ""}
              onChange={onSetUsername}
              disabled={loading}
            />
          </label>
        </div>
        <LoadingButton isSubmit isLoading={loading} type="success">
          Update profile
        </LoadingButton>
      </form>
    </div>
  );
}

const mapStateToProps = (state: AppState) => ({
  user: authSelectors.getUser(state) as User,
});

export default connect(mapStateToProps)(AccountEdit);
