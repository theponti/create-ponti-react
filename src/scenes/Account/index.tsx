import { connect } from 'react-redux';

import { AppState, authSelectors } from 'services/store';

type AccountProps = {
  user: User
};
function Account({ user }: AccountProps) {
  return (
    <div>
      <h1 className="text-3xl mb-8">My account</h1>
      <p>
        {user.name}
      </p>
      <p>
        <span className="font-bold text-lg mr-2">Email:</span>
        {user.email}
      </p>
    </div>
  );
}

const mapStateToProps = (state: AppState) => ({
  user: authSelectors.getUser(state) as User,
});

export default connect(mapStateToProps)(Account);
