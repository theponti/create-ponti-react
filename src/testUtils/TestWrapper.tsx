import PropTypes, { InferProps } from "prop-types";
import { Provider } from "react-redux";
import { BrowserRouter } from "react-router-dom";
import { store } from "services/store";

function TestWrapper({ children }: InferProps<typeof TestWrapper.propTypes>) {
  return (
    <BrowserRouter>
      <Provider store={store}>{children}</Provider>
    </BrowserRouter>
  );
}

TestWrapper.propTypes = {
  children: PropTypes.node.isRequired,
};

export default TestWrapper;
