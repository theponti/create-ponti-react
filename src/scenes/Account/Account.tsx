import AlertError from "components/Alerts/Error";
import LoadingButton from "components/Buttons/LoadingButton";
import PageWrap from "components/PageWrap";
import { useCallback } from "react";
import { connect } from "react-redux";
import { useNavigate } from "react-router-dom";
import { useDeleteUser } from "services/auth";

import { AppState, authSelectors } from "services/store";
import { supabase } from "services/supabase";
import AccountEdit from "./AccountEdit";

type AccountProps = {
  user: User;
};
function Account({ user }: AccountProps) {
  const navigate = useNavigate();
  const { deleteUser, error, loading } = useDeleteUser();
  const onDeleteAccount = useCallback(async () => {
    await deleteUser(user.id);
  }, [deleteUser, user]);
  const onLogoutClick = useCallback(async () => {
    await supabase.auth.signOut();
    navigate("/");
  }, [navigate]);

  return (
    <PageWrap>
      <h1 className="text-3xl mb-8">My account</h1>
      {error ? <AlertError message={error} /> : null}
      <div className="mb-8">
        <p>
          <span className="font-bold text-lg mr-2">Email:</span>
          {user.email}
        </p>
      </div>
      <div className="mb-8">
        <AccountEdit />
      </div>
      <hr />
      <button
        className="btn btn-outline mb-8 mt-8"
        data-testid="logoutButton"
        type="button"
        onClick={onLogoutClick}
        onKeyDown={onLogoutClick}
      >
        Logout
      </button>
      <div>
        <LoadingButton
          isLoading={loading}
          onClick={onDeleteAccount}
          type="error"
        >
          Delete account
        </LoadingButton>
      </div>
    </PageWrap>
  );
}

const mapStateToProps = (state: AppState) => ({
  user: authSelectors.getUser(state) as User,
});

export default connect(mapStateToProps)(Account);
