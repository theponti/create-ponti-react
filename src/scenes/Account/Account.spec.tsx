import { render } from '@testing-library/react';
import { describe, test } from 'vitest';

import { authSelectors } from 'services/store';
import TestWrapper from 'testUtils/TestWrapper';
import Account from '.';

const user = {
  id: 'foobar',
  email: 'foobar',
  name: 'foobar',
};

describe('Account', () => {
  beforeEach(() => {
    vi.spyOn(authSelectors, 'getUser').mockReturnValue(user);
  });

  test('should render', () => {
    const { container } = render(
      <TestWrapper>
        <Account />
      </TestWrapper>,
    );
    expect(container).toMatchSnapshot();
  });
});
