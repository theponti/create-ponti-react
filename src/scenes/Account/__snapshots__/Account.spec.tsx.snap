// Vitest Snapshot v1

exports[`Account > should be able to logout 1`] = `
<div>
  <div>
    <div>
      <h1>
        My account
      </h1>
      <p>
        <img
          alt="avatar"
          src="http://someimage"
        />
      </p>
      <p>
        foobar
      </p>
      <p>
        <b>
          Email verified:
        </b>
      </p>
      <p>
        <b>
          Email: 
        </b>
        example@example.com
      </p>
      <p>
        <button
          class=" css-1397b20"
          data-testid="logoutButton"
          type="button"
        >
          Log out
        </button>
      </p>
    </div>
  </div>
</div>
`;

exports[`Account > should render 1`] = `
<div>
  <div>
    <div />
  </div>
</div>
`;
